mne_sandbox/decoding/_time_frequency_csp.py

```diff
@@ -0,0 +1,255 @@
+            raise ValueError('sfreq must be a float or an int, got %s '
+                             'instead.' % type(n_cycles))
+        self.n_cycles = float(n_cycles)
+
+        # TODO:add test that the number of cycles is appropriate for epochs
+
+        self.estimator = estimator
+        self.scorer = scorer
+        self.n_jobs = n_jobs
+
+        # Assemble list of frequency range tuples
+        self._freq_ranges = zip(self.freqs[:-1], freqs[1:])
+
+        # Infer window spacing from the max freq and n cycles to avoid gaps
+        self._window_spacing = (n_cycles / np.max(freqs) / 2.)
+        self._centered_w_times = np.arange(tmin + self._window_spacing,
+                                           tmax - self._window_spacing,
+                                           self._window_spacing)[1:]  # noqa
+        self._n_windows = len(self._centered_w_times)
+
+    def _transform(self, epochs, y, w_tmin, w_tmax, method='', pos=[]):
+        """
+        Assumes data for one frequency band for one time window and fits CSP.
+        """
+        from sklearn import clone
+
+        # Crop data into time-window of interest
+        Xt = epochs.copy().crop(w_tmin, w_tmax).get_data()
+
+        # call fit or predict depending on calling function
+        if method == 'fit':
+            self.estimator = clone(self.estimator)
+            self._estimators[pos] = (self.estimator.fit(Xt, y))
+            return self
+
+        elif method == 'predict':
+            self.y_pred = self._estimators[pos].predict(Xt)
+
+        elif method == 'score':
+            self.y_pred = self._estimators[pos].predict(Xt)
+            return self._estimators[pos].score(Xt, y)
+
+    def fit(self, epochs, y=None):
+        """Train a classifier on each specified frequency and time slice.
+
+        Parameters
+        ----------
+        epochs : instance of Epochs
+            The epochs.
+        y : list or ndarray of int, shape (n_samples,) or None, optional
+            To-be-fitted model values. If None, y = epochs.events[:, 2].
+
+        Returns
+        -------
+        self : TimeFrequencyCSP
+            Returns fitted TimeFrequencyCSP object.
+        """
+        n_freq = len(self._freq_ranges)
+        n_window = len(self._centered_w_times)
+        self._scores = np.zeros([n_freq, n_window])
+        self._estimators = np.empty([n_freq, n_window], dtype=object)
+
+        if y is None:
+            y = epochs.events[:, 2]
+
+        # loop through each frequency range
+        for freq_ii, (fmin, fmax) in enumerate(self._freq_ranges):
+
+            # Infer window size based on the frequency being used
+            w_size = self.n_cycles / ((fmax + fmin) / 2.)  # in seconds
+
+            # filter the data at the desired frequency
+            epochs_bp = epochs.copy()
+            epochs_bp._data = filter_data(epochs_bp.get_data(), self.sfreq,
+                                          fmin, fmax, verbose='CRITICAL')
```

Comment:
A thought: maybe it is helpful to expose some parameters for the filtering? Perhaps not, as the number of parameters to this class will grow quite large.